README.md
# FlowCrypt: Encrypt Gmail with PGP


[FlowCrypt](https://flowcrypt.com/) Chrome plugin available at: https://chrome.google.com/webstore/detail/bnjglocicdkmhmoohhfkfkbbkejdhdgc/


20msgs

heavy+bg: 
12.7
12.7

heavy+local:
14.2
14.2

light+bg:
11.56
11.96
11.64

heavy+webworker:
13.54
13.50

light+bg+webworker:
11.63
11.63

light+bg+4webworkers:
9.31
8.49
8.81

light+bg+8webworkers:
8.37
7.78
7.75

light+bg+8workers+nodb:
8.69
7.60
7.75

light+bg+8workers+nodb+pruneapi:
8.68
8.59
8.55

light+bg+8workers+nodb+pruneapi+nostorage: !!! best about double speed / half exec time
7.07
7.10
7.06


light+bg+8workers+nodb+pruneapi+nostorage+zepto:
7.33
7.35
7.54

light+bg+8workers+nodb+pruneapi+nostorage+zepto+optimize$:
7.68
7.62
7.69

light+bg+8workers+nodb+pruneapi+nostorage+optimize$:
7.90
7.77
8.26

light+bg+8workers+nodb+pruneapi+nostorage+norequire:
7.62
7.67
7.70

light+bg+8workers+nodb+pruneapi+nostorage+norequire+zepto:
7.29
7.30
7.52




20 msgs - light+bg+nodb+pruneapi+nostorage
1 worker	9.31	9.20	9.17
8 workers	7.09	6.42	6.35

50msgs - light+bg+nodb+pruneapi+nostorage
1 worker	33.51	24.12	24.13
8 workers	15.25	15.44	15.41




random samples below (not definitive - tends to have large variance between runs)

----------------------- using jquery
0 | 0 | start
3 | 3 | anchorme.js
908 | 911 | jquery
26 | 937 | iso
9 | 946 | require.js
1 | 947 | lang
1 | 948 | factory
19 | 967 | mnemonic
2 | 969 | storage
66 | 1035 | common
4 | 1039 | pgp_block.js before db open
0 | 1039 | pgp_block.js after db open
19 | 1058 | pgp_block.js storage before
0 | 1058 | pgp_block.js storage after
1 | 1059 | pgp_block.js initialize
0 | 1059 | pgp_block.js decrypt and render start
2 | 1061 | pgp_block.js
0 | 1061 | end
3315 | 4376 | pgp_block.js decrypted
1 | 4377 | pgp_block.js rendering
699 | 5076 | pgp_block.js rendered


--------------------- using zepto: faster zepto load but longer overall execution
0 | 0 | start
3 | 3 | anchorme.js
5 | 8 | zepto
24 | 32 | iso
966 | 998 | require.js
1 | 999 | lang
1 | 1000 | factory
3 | 1003 | mnemonic
1 | 1004 | storage
43 | 1047 | common
6 | 1053 | pgp_block.js before db open
0 | 1053 | pgp_block.js after db open
21 | 1074 | pgp_block.js storage before
0 | 1074 | pgp_block.js storage after
0 | 1074 | pgp_block.js initialize
1 | 1075 | pgp_block.js decrypt and render start
2 | 1077 | pgp_block.js
0 | 1077 | end
4437 | 5514 | pgp_block.js decrypted
3 | 5517 | pgp_block.js rendering
139 | 5656 | pgp_block.js rendered

----------------- jquery + no require
0 | 0 | start
4 | 4 | anchorme.js
994 | 998 | jquery
2 | 1000 | iso
1 | 1001 | lang
1 | 1002 | factory
2 | 1004 | mnemonic
2 | 1006 | storage
21 | 1027 | common
5 | 1032 | pgp_block.js before db open
0 | 1032 | pgp_block.js after db open
3 | 1035 | pgp_block.js storage before
0 | 1035 | pgp_block.js storage after
0 | 1035 | pgp_block.js initialize
1 | 1036 | pgp_block.js decrypt and render start
1 | 1037 | pgp_block.js
1 | 1038 | end
4297 | 5335 | pgp_block.js decrypted
2 | 5337 | pgp_block.js rendering
494 | 5831 | pgp_block.js rendered

--------------------- zepto + no require
0 | 0 | start
3 | 3 | anchorme.js
5 | 8 | zepto
3 | 11 | iso
1 | 12 | lang
1 | 13 | factory
3 | 16 | mnemonic
2 | 18 | storage
564 | 582 | common
4 | 586 | pgp_block.js before db open
0 | 586 | pgp_block.js after db open
3 | 589 | pgp_block.js storage before
0 | 589 | pgp_block.js storage after
0 | 589 | pgp_block.js initialize
0 | 589 | pgp_block.js decrypt and render start
1 | 590 | pgp_block.js
0 | 590 | end
3829 | 4419 | pgp_block.js decrypted
3 | 4422 | pgp_block.js rendering
638 | 5060 | pgp_block.js rendered

chrome/changelog.txt
<div>

version 6.1.7 (stable only) on November 9, 2018: Reattest fix: https://github.com/FlowCrypt/flowcrypt-browser/issues/1328

version 6.1.6 on November 2, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/271?closed=1" target="_blank">Draft Newlines Fix</a>

version 6.1.5 on November 2, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/270?closed=1" target="_blank">Content Script Fix</a>

version 6.1.4 on November 1, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/269?closed=1" target="_blank">Attest Bugfix</a>

version 6.1.3 on October 31, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/268?closed=1" target="_blank">Experimental Module</a>

version 6.1.2 on October 29, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/267?closed=1" target="_blank">FlowCrypt Inbox</a>

version 6.1.1 on October 28, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/266?closed=1" target="_blank">Contacts page fix II</a>

version 6.1.0 on October 27, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/265?closed=1" target="_blank">Contacts page fix</a>

version 6.0.9 on October 27, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/264?closed=1" target="_blank">Refactoring</a>

version 6.0.8 on October 26, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/263?closed=1" target="_blank">HTML and Unicode</a>

version 6.0.7 on October 24, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/262?closed=1" target="_blank">Symantec Encryption Server Compatibility</a>

version 6.0.6 on October 23, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/261?closed=1" target="_blank">Reply button fix</a>

version 6.0.5 on October 19, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/260?closed=1" target="_blank">Special characters</a>

version 6.0.4 on October 14, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/259?closed=1" target="_blank">Google account changes</a>

version 6.0.3 on October 4, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/258?closed=1" target="_blank">Newlines</a>

version 6.0.2 on September 7, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/257?closed=1" target="_blank">Fix FlowCrypt Startup</a>

version 6.0.1 on September 5, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/256?closed=1" target="_blank">User id parse fix</a>

version 6.0.0 on August 26, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/255?closed=1" target="_blank">Mutt compatibility</a>

version 5.9.9 on August 16, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/254?closed=1" target="_blank">Skip invalid userids</a>

version 5.9.8 on August 16, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/253?closed=1" target="_blank">Import naked key</a>

version 5.9.7 on August 14, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/252?closed=1" target="_blank">Bugfixes</a>

version 5.9.6 on August 12, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/251?closed=1" target="_blank">Bugfixes</a>

version 5.9.5 on August 11, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/250?closed=1" target="_blank">Bugfixes</a>

version 5.9.4 on August 10, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/249?closed=1" target="_blank">Bugfixes</a>

version 5.9.3 on August 9, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/248?closed=1" target="_blank">Bugfixes</a>

version 5.9.2 on August 7, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/247?closed=1" target="_blank">Maintenance</a>

version 5.9.1 on August 7, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/246?closed=1" target="_blank">Error handling</a>

version 5.9.0 on August 7, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/245?closed=1" target="_blank">Attachment bugfix</a>

version 5.8.9 on August 6, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/244?closed=1" target="_blank">Attestation fix</a>

version 5.8.8 on August 6, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/243?closed=1" target="_blank">Pass phrase prompt</a>

version 5.8.7 on August 5, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/242?closed=1" target="_blank">Key decrypted error</a>

version 5.8.6 on August 3, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/241?closed=1" target="_blank">Address parsing fix</a>

version 5.8.5 on August 3, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/240?closed=1" target="_blank">Password mismatch fix</a>

version 5.8.4 on August 3, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/239?closed=1" target="_blank">Bugfixes</a>

version 5.8.3 on August 1, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/238?closed=1" target="_blank">Bugfixes</a>

version 5.8.2 on July 31, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/237?closed=1" target="_blank">Update Worker</a>

version 5.8.1 on July 31, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/236?closed=1" target="_blank">Firefox Fix</a>

version 5.8.0 on July 31, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/235?closed=1" target="_blank">OpenPGP.js v3</a>

version 5.7.2 on July 13, 2018: <a href="https://github.com/FlowCrypt/flowcrypt-browser/milestone/233?closed=1" target="_blank">Refactoring</a>


chrome/lib/iso-8859-2.js
/*! https://mths.be/iso-8859-2 v0.1.2 by @mathias | MIT license */
;(function(root) {

	// Detect free variables `exports`.
	var freeExports = typeof exports == 'object' && exports;

	// Detect free variable `module`.
	var freeModule = typeof module == 'object' && module &&
		module.exports == freeExports && module;

	// Detect free variable `global`, from Node.js/io.js or Browserified code,
	// and use it as `root`.
	var freeGlobal = typeof global == 'object' && global;
	if (freeGlobal.global === freeGlobal || freeGlobal.window === freeGlobal) {
		root = freeGlobal;
	}

	/*--------------------------------------------------------------------------*/

	var object = {};
	var hasOwnProperty = object.hasOwnProperty;
	var stringFromCharCode = String.fromCharCode;

	var INDEX_BY_CODE_POINT = {'128':0,'129':1,'130':2,'131':3,'132':4,'133':5,'134':6,'135':7,'136':8,'137':9,'138':10,'139':11,'140':12,'141':13,'142':14,'143':15,'144':16,'145':17,'146':18,'147':19,'148':20,'149':21,'150':22,'151':23,'152':24,'153':25,'154':26,'155':27,'156':28,'157':29,'158':30,'159':31,'160':32,'164':36,'167':39,'168':40,'173':45,'176':48,'180':52,'184':56,'193':65,'194':66,'196':68,'199':71,'201':73,'203':75,'205':77,'206':78,'211':83,'212':84,'214':86,'215':87,'218':90,'220':92,'221':93,'223':95,'225':97,'226':98,'228':100,'231':103,'233':105,'235':107,'237':109,'238':110,'243':115,'244':116,'246':118,'247':119,'250':122,'252':124,'253':125,'258':67,'259':99,'260':33,'261':49,'262':70,'263':102,'268':72,'269':104,'270':79,'271':111,'272':80,'273':112,'280':74,'281':106,'282':76,'283':108,'313':69,'314':101,'317':37,'318':53,'321':35,'322':51,'323':81,'324':113,'327':82,'328':114,'336':85,'337':117,'340':64,'341':96,'344':88,'345':120,'346':38,'347':54,'350':42,'351':58,'352':41,'353':57,'354':94,'355':126,'356':43,'357':59,'366':89,'367':121,'368':91,'369':123,'377':44,'378':60,'379':47,'380':63,'381':46,'382':62,'711':55,'728':34,'729':127,'731':50,'733':61};
	var INDEX_BY_POINTER = {'0':'\x80','1':'\x81','2':'\x82','3':'\x83','4':'\x84','5':'\x85','6':'\x86','7':'\x87','8':'\x88','9':'\x89','10':'\x8A','11':'\x8B','12':'\x8C','13':'\x8D','14':'\x8E','15':'\x8F','16':'\x90','17':'\x91','18':'\x92','19':'\x93','20':'\x94','21':'\x95','22':'\x96','23':'\x97','24':'\x98','25':'\x99','26':'\x9A','27':'\x9B','28':'\x9C','29':'\x9D','30':'\x9E','31':'\x9F','32':'\xA0','33':'\u0104','34':'\u02D8','35':'\u0141','36':'\xA4','37':'\u013D','38':'\u015A','39':'\xA7','40':'\xA8','41':'\u0160','42':'\u015E','43':'\u0164','44':'\u0179','45':'\xAD','46':'\u017D','47':'\u017B','48':'\xB0','49':'\u0105','50':'\u02DB','51':'\u0142','52':'\xB4','53':'\u013E','54':'\u015B','55':'\u02C7','56':'\xB8','57':'\u0161','58':'\u015F','59':'\u0165','60':'\u017A','61':'\u02DD','62':'\u017E','63':'\u017C','64':'\u0154','65':'\xC1','66':'\xC2','67':'\u0102','68':'\xC4','69':'\u0139','70':'\u0106','71':'\xC7','72':'\u010C','73':'\xC9','74':'\u0118','75':'\xCB','76':'\u011A','77':'\xCD','78':'\xCE','79':'\u010E','80':'\u0110','81':'\u0143','82':'\u0147','83':'\xD3','84':'\xD4','85':'\u0150','86':'\xD6','87':'\xD7','88':'\u0158','89':'\u016E','90':'\xDA','91':'\u0170','92':'\xDC','93':'\xDD','94':'\u0162','95':'\xDF','96':'\u0155','97':'\xE1','98':'\xE2','99':'\u0103','100':'\xE4','101':'\u013A','102':'\u0107','103':'\xE7','104':'\u010D','105':'\xE9','106':'\u0119','107':'\xEB','108':'\u011B','109':'\xED','110':'\xEE','111':'\u010F','112':'\u0111','113':'\u0144','114':'\u0148','115':'\xF3','116':'\xF4','117':'\u0151','118':'\xF6','119':'\xF7','120':'\u0159','121':'\u016F','122':'\xFA','123':'\u0171','124':'\xFC','125':'\xFD','126':'\u0163','127':'\u02D9'};

	// https://encoding.spec.whatwg.org/#error-mode
	var error = function(codePoint, mode) {
		if (mode == 'replacement') {
			return '\uFFFD';
		}
		if (codePoint != null && mode == 'html') {
			return '&#' + codePoint + ';';
		}
		// Else, `mode == 'fatal'`.
		throw Error();
	};

	// https://encoding.spec.whatwg.org/#single-byte-decoder
	var decode = function(input, options) {
		var mode;
		if (options && options.mode) {
			mode = options.mode.toLowerCase();
		}
		// “An error mode […] is either `replacement` (default) or `fatal` for a
		// decoder.”
		if (mode != 'replacement' && mode != 'fatal') {
			mode = 'replacement';
		}
		var length = input.length;
		var index = -1;
		var byteValue;
		var pointer;
		var result = '';
		while (++index < length) {
			byteValue = input.charCodeAt(index);
			// “If `byte` is in the range `0x00` to `0x7F`, return a code point whose
			// value is `byte`.”
			if (byteValue >= 0x00 && byteValue <= 0x7F) {
				result += stringFromCharCode(byteValue);
				continue;
			}
			// “Let `code point` be the index code point for `byte − 0x80` in index
			// `single-byte`.”
			pointer = byteValue - 0x80;
			if (hasOwnProperty.call(INDEX_BY_POINTER, pointer)) {
				// “Return a code point whose value is `code point`.”
				result += INDEX_BY_POINTER[pointer];
			} else {
				// “If `code point` is `null`, return `error`.”
				result += error(null, mode);
			}
		}
		return result;
	};

	// https://encoding.spec.whatwg.org/#single-byte-encoder
	var encode = function(input, options) {
		var mode;
		if (options && options.mode) {
			mode = options.mode.toLowerCase();
		}
		// “An error mode […] is either `fatal` (default) or `HTML` for an
		// encoder.”
		if (mode != 'fatal' && mode != 'html') {
			mode = 'fatal';
		}
		var length = input.length;
		var index = -1;
		var codePoint;
		var pointer;
		var result = '';
		while (++index < length) {
			codePoint = input.charCodeAt(index);
			// “If `code point` is in the range U+0000 to U+007F, return a byte whose
			// value is `code point`.”
			if (codePoint >= 0x00 && codePoint <= 0x7F) {
				result += stringFromCharCode(codePoint);
				continue;
			}
			// “Let `pointer` be the index pointer for `code point` in index
			// `single-byte`.”
			if (hasOwnProperty.call(INDEX_BY_CODE_POINT, codePoint)) {
				pointer = INDEX_BY_CODE_POINT[codePoint];
				// “Return a byte whose value is `pointer + 0x80`.”
				result += stringFromCharCode(pointer + 0x80);
			} else {
				// “If `pointer` is `null`, return `error` with `code point`.”
				result += error(codePoint, mode);
			}
		}
		return result;
	};

	var iso88592 = {
		'encode': encode,
		'decode': decode,
		'labels': [
			'csisolatin2',
			'iso-8859-2',
			'iso-ir-101',
			'iso8859-2',
			'iso88592',
			'iso_8859-2',
			'iso_8859-2:1987',
			'l2',
			'latin2'
		],
		'version': '0.1.2'
	};

	// Some AMD build optimizers, like r.js, check for specific condition patterns
	// like the following:
	if (
		typeof define == 'function' &&
		typeof define.amd == 'object' &&
		define.amd
	) {
		define(function() {
			return iso88592;
		});
	}	else if (freeExports && !freeExports.nodeType) {
		if (freeModule) { // in Node.js, io.js or RingoJS v0.8.0+
			freeModule.exports = iso88592;
		} else { // in Narwhal or RingoJS v0.7.0-
			for (var key in iso88592) {
				iso88592.hasOwnProperty(key) && (freeExports[key] = iso88592[key]);
			}
		}
	} else { // in Rhino or a web browser
		root.iso88592 = iso88592;
	}

}(this));

if(window.flowcrypt_profile) {
  window.flowcrypt_profile.add('iso');
}